UV Migration (UV, Ruff, Pandas v2)

DavidGeorge528 opened the pull request:
This PR migrates the project to UV away from Poetry. The key changes are:

1. Poetry removed as project manager, replaced with UV
2. Black and iSort removed as formatters, replaced with Ruff
3. Python 3.8 support removed as its EOL

All code is updated to reflect the above changes, including:
1. pyproject.toml (Removed poetry, black, isort, added UV and Ruff
2. Makefile (Uses UV and Ruff)
3. tox (Removed 3.8, uses UV for faster running (~30 seconds to run whole suite))
4. pre-commit just uses Ruff to ensure linting
5. Pandas updated to v2 and made a core dependency
6. Numpy added with minimum version numpy>1.24.4 (Resolves #99)


Bug fixes:
1. `__eq__` added to `Category` to fix recursive equality (Note: Might be worth moving it to `BaseModel`)
2. Some typing issues fixed

All tests run and tox completes with no errors

isaacharrisholt commented on .github/workflows/tests.yml:
```diff
@@ -10,7 +10,7 @@ jobs:
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
-        python-version: ['3.8', '3.9', '3.10', '3.11']
+        python-version: ['3.9', '3.10', '3.11']
```
If we're changing things, might as well test 3.12 and 3.13 too

DavidGeorge528 replied:
Added py3.12 to the GitHub actions, but py3.13 was giving me errors with Numpy, on Mac at least. Also it required an update to numpy 2.1

isaacharrisholt replied:
Can we not update numpy now that we're updating pandas?

In fact, do we even need to have numpy as a dependency anymore? I can't remember if it's used elsewhere

DavidGeorge528 replied:
You're correct, since the move the Pandas 2, the numpy minimum version is above that set in the pyproject so the issue in #91  is fixed by default. Removed the dedicated numpy dependency (It's still needed by pandas and pydantic)

isaacharrisholt commented on makefile:
```diff
@@ -1,25 +1,25 @@
+	( uv run ruff check $(PKG_NAME); uv run ruff check tests )
 
-.PHONY:		wheel
+.PHONY:	wheel
```
This fails when I run it. Are we missing `wheel`?

DavidGeorge528 replied:
Fixed, was still using poetry syntax, UVs command is slightly differnt

isaacharrisholt commented on quiffen/core/category.py:
```diff
@@ -511,27 +516,25 @@ def add_categories_to_container(
     # Add categories in hierarchy
-    categories_is_dict = isinstance(categories, dict)
     if new_category.hierarchy != new_category.name or new_category.children:
         # Get the root category in the chain
         root = new_category.traverse_up()[-1]
 
-        iterator = categories.values() if categories_is_dict else categories
+        iterator = categories.values() if isinstance(categories, dict) else categories
```
I assume these changes are for type-checking purposes?

DavidGeorge528 replied:
Yeah, can revert if you'd rather, mypy doesn't follow the logic of a flag dictating somethings type so was complaining

isaacharrisholt replied:
We shouldn't be using mypy - the type checker for quiffen is pyright. I'm not opposed to this, but it's _technically_ less efficient than what we had previously

isaacharrisholt commented on quiffen/core/qif.py:
```diff
@@ -17,13 +18,6 @@
 from quiffen.core.security import Security
 from quiffen.core.transaction import Transaction
 
-try:
-    import pandas as pd
-
-    PANDAS_INSTALLED = True
-except ModuleNotFoundError:
-    PANDAS_INSTALLED = False
-
```
Revert this. Pandas is an optional dependency.

DavidGeorge528 replied:
Reverted, moved the import inside the function with a helpful error message. Pandas is not installed by default now, `uv sync --all-extras` or `uv sync --extra pandas` is required to install pandas.

isaacharrisholt commented on tests/test_base.py:
```diff
@@ -9,7 +9,7 @@ def test_create_field():
     assert field.line_code == "T"
     assert field.attr == "test"
-    assert field.type == str
+    assert field.type is str
```
Why this change?

boxydog replied:
This is a ruff rule. https://docs.astral.sh/ruff/rules/type-comparison/#type-comparison-e721

isaacharrisholt commented on tests/test_qif.py:
```diff
@@ -634,7 +634,6 @@ def test_get_data_dicts_accounts():
 
     assert len(data_dicts) == 1
     expected = account.to_dict()
-    del expected["_last_header"]
```
Was this causing problems?

DavidGeorge528 replied:
I was getting a `KeyError`:

```FAILED tests/test_qif.py::test_get_data_dicts_accounts - KeyError: '_last_header'```

Since by default pydantic doesn't treat properties starting with an underscore as fields, so they don't get included in the dict method

This also brings up a much bigger refactor, which is migrating to pydantic V2. Several methods (e.g.`dict()`) are deprecated even in V1.

DavidGeorge528 replied:
Ignore my previous reply, I'd accidentally set pydantic>2 rather than <2 in the deps
I've put the line back in and fixed the dependency issue.

Worth noting, bare minimum update to pydantic>2 is removing that line from the test file, everything else seems to run with legacy support (Lots of deprecated methods)

isaacharrisholt replied:
Yep, Pydantic V2 would be another great update at some point

isaacharrisholt commented on tox.ini:
```diff
@@ -1,47 +1,39 @@
 requires =
     tox>=4
-envlist = py{38,39,310,311,312}, black, ruff, pyright, isort
+envlist = py{39,310,311,312}, ruff, pyright
```
Again, we should look at 3.13 now too

DavidGeorge528 replied:
Similar to above, tried to add 3.13 but had an issue with numpy not installing during a `uv sync`

DavidGeorge528 replied:
I've managed to get python 3.13 working if I pin numpy as a requirement with v2 as a minimum. What do you think?

isaacharrisholt replied:
I'm 95% sure we don't need numpy as a dependency at all, right? It's only pandas that depends on it. Can we add it as a dev dep perhaps? Or maybe part of the `pandas` group of optionals?

DavidGeorge528 replied:
It's not a direct dependency no, and its been removed from the dependency list, but for python 3.13 without a strict >2 dependency then it fails to install an older version

isaacharrisholt commented on pyproject.toml:
```diff
@@ -1,56 +1,90 @@
+[project]
+authors = [
+    {name = "Isaac Harris-Holt", email = "isaac@harris-holt.com"},
+]
+license = {text = "GPL-3.0-or-later"}
+requires-python = "<4.0,>=3.9"
+dependencies = [
+    "pydantic>2.0.0,>=1.10.2",
+    "pandas>2.2.0,<3.0.0",
+    "numpy>1.24.4,<2.0.0",
+    "python-dateutil>=2.8.2,<3.0.0",
+]
 name = "quiffen"
-version = "2.0.13"
+version = "2.1.0"
```
Revert this - I've got a few PRs to look into then I'll bump the version myself.

DavidGeorge528 replied:
Reverted

isaacharrisholt reviewed (changes requested):
Thank you! This is great work overall - super useful stuff.

DavidGeorge528 commented:
All comments resolved to the best of my abilities. A few notes summarised below:

1. Python 3.13 seems to have incompatibility with the current numpy version, and I even had issues with numpy>2.1, so omitted it from this PR for now
2. Pandas moved to optional dependency, can be installed with `uv sync --all-extras` or `uv sync --extra pandas`. The import was moved into the function to keep mypy happy (TYPE_CHECKING import of pandas used to maintain type hints in the return type)
3. Version number bump reverted
4. Fixed issue with makefile for wheel
5. Brought final changes from #102 to make github actions run pre-commit

Tox runs locally with no issues
GitHub actions runs on my personal fork of the PR with no issues

Future TODOs:

1. Add support for python 3.13
2. Update pydantic to v2 (Bare minimum required is removing `del expected["_last_header"]` from `tests/test_qif.py` line 637, however, this results in many deprecation warnings)

isaacharrisholt commented on quiffen/core/category.py:
```diff
@@ -470,7 +475,7 @@ def from_list(cls, lst: List[str]) -> Category:
 
 
-ListOrDictOfCategories = Union[List[Category], Dict[str, Category]]
+T = TypeVar("T", List[Category], Dict[str, Category])
```
Happy to make this a `TypeVar`, but if we could use better naming that would be great. `T` means nothing in this context.

DavidGeorge528 replied:
Naming for `TypeVar`s, defined [here](https://peps.python.org/pep-0008/#type-variable-names), are meant to be short, and typically `T` is used. I can change it to something like `LDC` or just `C` or something else, but it wont make it much more readable.

isaacharrisholt replied:
See https://github.com/isaacharrisholt/quiffen/pull/105#issuecomment-2613956764

isaacharrisholt commented on quiffen/core/qif.py:
```diff
@@ -576,10 +572,12 @@ def to_dataframe(
-            )
+        try:
+            import pandas as pd
+        except ImportError as e:
+            raise ImportError(
+                "The pandas package is required to convert Qif objects to DataFrames. Please install it using `pip install quiffen[pandas]`."
+            ) from e
```
I believe `ModuleNotFoundError` was correct here

DavidGeorge528 replied:
Fixed

DavidGeorge528 commented:
Hey, is there any movement on getting this PR merged? Are there any other changes you need? Thanks

isaacharrisholt commented:
Hey, sorry - I've been without internet for the last week due to a fire. I've finally managed to get a dongle to use my PC with a hotspot. I'd still prefer the type var gets a name like `CategoryContainer` or something that gives a but more context and is less likely to conflict with stuff in the future, but otherwise this is great.

DavidGeorge528 commented:
Oh sorry to hear! Hope everyone is alright!

I can change the TypeVar if you'd rather, even though its against conventions. What about the other TypeVar in `core/base.py`?

isaacharrisholt commented:
All good, thank you. Just took out our internet and some lights.

And it's fine to ignore convention if it makes code more readable. Would be good to apply the same logic to both.

DavidGeorge528 commented:
I have updated the `TypeVar` that I added, however, after looking at the other `TypeVar` in `core/base.py` I think the implementation is incorrect

The class `BaseModel` is setup as follows:

```python
T = TypeVar("T")
class BaseModel(PydanticBaseModel, Generic[T]):
    ...
```

This means that either:

- Subclasses define `T` (`class Subclass(BaseModel[SomeType]):`)
- Instantiated objects define `T` (`a = Subclass[SomeType]()`)

Neither of these are done, and the `TypeVar` is never actually instantiated to any type throughout the codebase. The subclasses just overwrite the type with their class. I believe, to a type checker the current implementation has the base superclass return type for `from_list` as `Unknown` and all subclasses have it defined. This can actually be observed in an IDE by looking at the `from_string` signature, which has not been overridden by the `Category` class, and as result the type of its return is `Unknown`

<img width="408" alt="image" src="https://github.com/user-attachments/assets/25264f4e-e36d-4f79-988b-cf303ea969e9" />
<img width="382" alt="image" src="https://github.com/user-attachments/assets/1a765a2e-d817-433d-89d6-28e0d82974d1" />



I believe the actual implementation should be:

```python
class BaseModel(PydanticBaseModel):
    @classmethod
    @abstractmethod
    def from_list(cls, lst: List[str]) -> BaseModel:
        pass
```

Then all subclasses change `BaseModel` to themselves. However, this also has a problem, as the `from_list` implementation has the return type `Tuple[Transaction, Dict[str, Class]]`, which does not fit the superclass abstract implementation, and it never will.

There may also be a way of making the `TypeVar` work, but its not working as intended in the current implementation. As result I've not changed it from `T`.

isaacharrisholt commented:
I think the correct implementation would actually be to bind `T` to `BaseClass`, or, from Python 3.11, to return `Self`. I believe you can use `typing_extensions` to get `Self` in older Python versions, but I don't want to have to deal with the hassle of having different behaviour per version. I'll add this myself in a separate PR though.
